mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs

```diff
@@ -1710,6 +1905,432 @@ public override string ToString ()
 		}
 	}
 
+	// Base class for all Sockaddr types.
+	// This class is not abstract, instances of this class can be used to determine the sa_family value.
+	// This class and all classes which are deriving from it and are passed to the native code have to be blittable.
+	[CLSCompliant (false)]
+	[StructLayout (LayoutKind.Sequential)]
+	public class Sockaddr {
+		internal SockaddrType type;
+		public UnixAddressFamily sa_family;
+
+		public Sockaddr ()
+		{
+			this.type = SockaddrType.Sockaddr;
+			this.sa_family = UnixAddressFamily.AF_UNSPEC;
+		}
+		internal Sockaddr (SockaddrType type, UnixAddressFamily sa_family)
+		{
+			this.type = type;
+			this.sa_family = sa_family;
+		}
+
+		[DllImport (Syscall.MPH, SetLastError=true, 
+				EntryPoint="Mono_Posix_Sockaddr_GetNativeSize")]
+		static extern int GetNativeSize (ref _SockaddrDynamic address, out long size);
+		[DllImport (Syscall.MPH, SetLastError=true, 
+				EntryPoint="Mono_Posix_Sockaddr_GetNativeSize")]
+		static extern int GetNativeSize (Sockaddr address, out long size);
+		public unsafe long GetNativeSize ()
+		{
+			long size;
+			if (this != null && this.IsDynamic) {
+				fixed (byte* data = this.DynamicData ()) {
+					_SockaddrDynamic dyn = new _SockaddrDynamic (this, data, false);
+					if (GetNativeSize (ref dyn, out size) != 0)
+						throw new ArgumentException ("this", "Failed to get size of native struct");
+				}
+			} else {
+				if (GetNativeSize (this, out size) != 0)
+					throw new ArgumentException ("this", "Failed to get size of native struct");
+			}
+			return size;
+		}
+
+		internal bool IsDynamic {
```

Comment:
> I'm not sure what IsDynamic is supposed to mean, though I assume it has something to do with SockaddrType.MustBeWrapped:

It indicates whether the `Sockaddr` has to be wrapped using `_DynamicSockaddr` (i.e. whether the `SockaddrType.MustBeWrapped` flag is set).



> Perhaps IsWrapped would be better? Though I'm not sure of that name either. It would provide consistency with the MustBeWrapped flag...

I'll rename it to `NeedsWrapper`. (`IsWrapped` is not really correct.)

Shouldn't be too important though, `NeedsWrapper` also is internal.